provider/datagen/src/transform/cldr/datetime/neo.rs

```diff
@@ -383,11 +462,156 @@ fn months_convert(
     }
 }
 
+fn datepattern_convert(
+    data: &ca::Dates,
+    length: PatternLength,
+    _hc: Option<CoarseHourCycle>,
+) -> Result<DatePatternV1<'static>, DataError> {
+    let pattern = data.date_formats.get_pattern(length);
+
+    let pattern = pattern
+        .get_pattern()
+        .parse()
+        .expect("failed to parse pattern");
+    Ok(DatePatternV1 { pattern })
+}
+
+fn datetimepattern_convert(
+    data: &ca::Dates,
+    length: PatternLength,
+    _hc: Option<CoarseHourCycle>,
+) -> Result<DateTimePatternV1<'static>, DataError> {
+    let pattern = data.datetime_formats.get_pattern(length);
+
+    let pattern = pattern
+        .get_pattern()
+        .parse()
+        .expect("failed to parse pattern");
+    Ok(DateTimePatternV1 { pattern })
+}
+
+fn timepattern_convert(
+    data: &ca::Dates,
+    length: PatternLength,
+    hc: Option<CoarseHourCycle>,
+) -> Result<TimePatternV1<'static>, DataError> {
+    let pattern = data.time_formats.get_pattern(length);
+
+    let pattern_str = pattern.get_pattern();
+    let pattern = pattern_str.parse().expect("failed to parse pattern");
+    // We only get an hc if we're generating the non-default hc, so we know we must replace it in the pattern
+    let pattern: pattern::runtime::Pattern = if let Some(hc) = hc {
+        let length_combinations_v1 = GenericLengthPatternsV1::from(&data.datetime_formats);
+        let skeletons_v1 = DateSkeletonPatternsV1::from(data);
+        let pattern =
+            hc.apply_on_pattern(&length_combinations_v1, &skeletons_v1, pattern_str, pattern);
+        let pattern = pattern
+            .as_ref()
+            .expect("Failed to apply a coarse hour cycle to a full pattern.");
```

Comment:
Observation: it's good if this assertion succeeds for all ICU4X locales, but I don't know if it's perpetually guaranteed. `hc.apply_on_pattern` performs a skeleton resolution algorithm, which should normally succeed but I don't know if it's guaranteed to succeed if sufficient data is not present (perhaps for sub-basic locales).

Reply:
Yeah I'm not 100% sure of this either. Might be worth doing another pass on the panics here later.